packages/airnode-node/src/handlers/start-coordinator.ts

```diff
@@ -5,100 +5,176 @@ import * as logger from '../logger';
+  const [logs, processedAggregatedApiCalls] = await calls.callApis(
+    aggregatedApiCalls,
+    logOptions,
+    getWorkerOptions(state)
   );
-  if (hasNoRequests) {
-    logger.info('No actionable requests detected. Returning...', logOptions);
-    return state2;
-  }
+  logger.logPending(logs, logOptions);
+
+  logger.info('Executed API calls', logOptions);
+  const processedAggregatedApiCallsById = keyBy(processedAggregatedApiCalls, 'id');
+  return coordinatorState.update(state, { aggregatedApiCallsById: processedAggregatedApiCallsById });
+}
 
+function disaggregateApiCalls(state: CoordinatorState) {
+  const { config, coordinatorId } = state;
+  const logOptions = logger.buildBaseOptions(config, { coordinatorId });
+
+  // TODO: Disaggregation should be chain agnostic - currently only updated EVM providers are returned
+  const [logs, evmProvidersWithApiResponses] = calls.disaggregate(state);
+  logger.logPending(logs, logOptions);
+
+  logger.info('Disaggregated API calls', logOptions);
+  return coordinatorState.update(state, { providerStates: { evm: evmProvidersWithApiResponses } });
+}
+
+async function initiateTransactions(state: CoordinatorState) {
+  const { providerStates, config, coordinatorId } = state;
+  const logOptions = logger.buildBaseOptions(config, { coordinatorId });
+
+  providerStates.evm.map(async (evmProviderState) => {
+    logger.info(`Forking to submit transactions for EVM provider:${evmProviderState.settings.name}`, logOptions);
+  });
+  const [logs, processedProviders] = await providers.processRequests(providerStates, getWorkerOptions(state));
+  logger.logPending(logs, logOptions);
+
+  logger.info('Forking to submit transactions complete', logOptions);
+  return coordinatorState.update(state, { providerStates: processedProviders });
+}
+
+function applyChainRequestLimits(state: CoordinatorState) {
+  const { config, coordinatorId, providerStates } = state;
+  const logOptions = logger.buildBaseOptions(config, { coordinatorId });
+
+  logger.info('Applying chain request limits', logOptions);
+  const [logs, processedProviders] = calls.applyChainLimits(config, providerStates);
+  logger.logPending(logs, logOptions);
+
+  logger.info('Applied chain request limits', logOptions);
+  return coordinatorState.update(state, { providerStates: processedProviders });
+}
+
+async function coordinator(config: Config): Promise<CoordinatorState> {
   // =================================================================
-  // STEP 3: Group API calls with respect to request IDs
+  // STEP 1: Create a blank coordinator state
   // =================================================================
-  const flatApiCalls = flatMap(state2.providerStates.evm, (provider) => provider.requests.apiCalls);
-  const aggregatedApiCallsById = calls.aggregate(state2.config, flatApiCalls);
-  const state3 = state.update(state2, { aggregatedApiCallsById });
+  let state = createInitialCoordinatorState(config);
```

Comment:
I assume this is much better for memory usage? I'm not sure exactly how JS garbage collects things like `state1` after they're no longer needed

Reply:
It doesn't matter for performance, but it's a bit more readable and less error prone. In the past we had a potential bug where we had code like this:
```
const state5 = {...state3, ...someOtherThings}
```
where we should use state4 instead. Now there is no way for this bug to occur.